### PR TITLE
net/pkt: correct PF_PACKET family sending errors

### PR DESCRIPTION
**Pull request by zhhyu7**
## Summary
Enable the pkt_sendmsg interface to send packets containing Layer 2 headers.
## Testing
qemu x64

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by acassis** [comment]
@zhhyu7 I didn't follow: you added support to negative offset, but then you check if it is negative and for it to 0. So why don't stay with unsigned implementation.

My first thought was that negative offset could introduce vulnerabilities, but then I saw you are checking for negative value. I'm confuse

**Comment by zhhyu7** [comment]
> @zhhyu7 I didn't follow: you added support to negative offset, but then you check if it is negative and for it to 0. So why don't stay with unsigned implementation.
> 
> My first thought was that negative offset could introduce vulnerabilities, but then I saw you are checking for negative value. I'm confuse...
> 
> Please explain the issue that this modification solves.

Hi @acassis, The ```socket(PF_PACKET, SOCK_RAW)``` may send Layer 2 header information filled under user space, at present, this part of information cannot be copied to the space reserved for ```LL_GUARD_SIZE``` by ```io_data``` through ```devif_send```, even though iob_copyin supports negative offset. I don't have a better idea yet, is there a better modification suggestion for this situation? Thank you.

**Comment by acassis** [comment]
@zhhyu7 In fact iob_copyin_internal already accept negative offset and check for its range to avoid issues:

```
  if ((int)(offset - iob->io_pktlen) > 0)
    {
      ioberr("ERROR: offset is past the end of data: %d > %u\n",
             offset, iob->io_pktlen);
      return -ESPIPE;
    }

  if ((int)(offset + iob->io_offset) < 0)
    {
      ioberr("ERROR: offset is before the start of data: %d < %d\n",
             offset, -(int)iob->io_offset);
      return -ESPIPE;
    }
```

So, ignore my warnings

File: net/pkt/pkt_sendmsg.c [hunk 1]
```diff
@@ -106,13 +106,14 @@ static uint16_t psock_send_eventhandler(FAR struct net_driver_s *dev,
...
           int ret = devif_send(dev, pstate->snd_buffer,
-                               pstate->snd_buflen, 0);
+                               pstate->snd_buflen, -NET_LL_HDRLEN(dev));
           if (ret <= 0)
             {
               pstate->snd_sent = ret;
               goto end_wait;
             }
 
+          dev->d_len       = dev->d_sndlen -= NET_LL_HDRLEN(dev);
```
Comment (pkarashchenko) [inline net/pkt/pkt_sendmsg.c:116]: Can we spit this to multiple lines for readability?

Comment (acassis) [reply net/pkt/pkt_sendmsg.c:116]: Good point @pkarashchenko this code is confuse!

**Pull request merged**